inginious/frontend/pages/course_admin/student_list.py

```diff
@@ -192,33 +198,48 @@ def post_audiences(self, course, data, active_tab, msg, error):
+                # Furthermore, FileStorage.stream seems to inherit ʻio.BufferedIOBase`, so this stream should be boiled.
+                reader = csv.reader(io.TextIOWrapper(data["audiencefile"], encoding='utf-8'))
+                csv_data = list(reader)  # read everything else into a list of rows
+                auth_method_ids = self.user_manager.get_auth_methods().keys()
+                for line in csv_data:
+                    if len(line) != 4:
+                        msg["audiences"] = _("File wrongly formatted.")
+                        error["audiences"] = True
+                if "audiences" not in error or not error["audiences"]:
+                    stud_list, aud_li, oth_stu, u_info = self.get_user_lists(course)
                     self.database.audiences.delete_many({"courseid": course.get_id()})
-                    audiences = custom_yaml.load(data["audiencefile"].read())
-                else:
-                    audiences = json.loads(data["audiences"])
-
-                for index, new_audience in enumerate(audiences):
-                    # In case of file upload, no id specified
-                    new_audience['_id'] = new_audience['_id'] if '_id' in new_audience else 'None'
-
-                    # Update the audience
-                    audience, errors = self.update_audience(course, new_audience['_id'], new_audience)
-
-                    # If file upload was done, get the default audience id
-                    audienceid = audience['_id']
-                    errored_students += errors
-
-                if len(errored_students) > 0:
-                    msg["audiences"] = _("Changes couldn't be applied for following students :") + "<ul>"
-                    for student in errored_students:
-                        msg["audiences"] += "<li>" + student + "</li>"
-                    msg["audiences"] += "</ul>"
-                    error["audiences"] = True
-                elif not error:
-                    msg["audiences"] = _("Audience updated.")
+                    for user_id, field, role, description in csv_data:
+                        user_id = user_id.strip()
+                        field = field.strip()
+                        role = role.strip()
+                        description = description.strip()
+                        if field not in ["username", "email", "realname"] + list(auth_method_ids) \
+                                or role not in ["student", "tutors"]:
+                            continue
+                        if field != "username":
+                            user = self.database.users.find_one({field: user_id})
+                            user_id = user["username"] if user is not None else ""
+                            # TODO - Handle user not found
+                        query = {"courseid": course.get_id(), "description": description}
+
+                        existing_audience = self.database.audiences.find_one(query)
```

Comment:
You shouldn't perform such a merge (and especially mongo queries) for each student in the file. Parse the file, create the audiences dictionaries and then only update the existing audiences if needed. This way you keep the amount of queries to the minimum (or at least, a reasonable amount).